lms/djangoapps/instructor/views/api.py

```diff
@@ -2707,10 +2707,141 @@ def start_certificate_regeneration(request, course_id):
+    :return: CertificateWhitelist item in dict format containing all the keys required by
+             backbone model CertificateException.
+    """
+    if CertificateWhitelist.objects.filter(user=student, course_id=course_key, whitelist=True).count() > 0:
+        raise ValueError(
+            _("Student (username/email={user}) already in certificate exception list.").format(user=student.username)
+        )
+
+    certificate_white_list, __ = CertificateWhitelist.objects.get_or_create(
+        user=student,
+        course_id=course_key,
+        defaults={
+            'whitelist': True,
+            'notes': certificate_exception.get('notes', '')
+        }
+    )
+
+    exception = dict({
+        'id': certificate_white_list.id,
+        'user_email': student.email,
+        'user_name': student.username,
+        'user_id': student.id,
+        'created': certificate_white_list.created.strftime("%A, %B %d, %Y"),
+    })
+
+    return exception
+
+
+def remove_certificate_exception(course_key, student):
+    """
+    Remove certificate exception for given course and student from CertificateWhitelist table and
+    invalidate its GeneratedCertificate if present.
+
+    :param course_key: course key of the course for whom to delete certificate.
+    :param student: User object for whom to add Certificate Exception
+    :return:
+    """
+    try:
+        certificate_exception = CertificateWhitelist.objects.get(user=student, course_id=course_key)
+    except ObjectDoesNotExist:
+        raise ValueError(
+            _('Certificate exception [user={}] does not exist in '
+              'certificate white list.').format(student.username)
+        )
+
+    try:
+        generated_certificate = GeneratedCertificate.objects.get(user=student, course_id=course_key)
+        generated_certificate.invalidate()
+    except ObjectDoesNotExist:
+        # Certificate has not been generated yet, so just remove the certificate exception from white list
+        pass
+    certificate_exception.delete()
+
+
+def parse_request_data(request):
+    """
+        Parse request data into Certificate Exception and User object.
+        Certificate Exception is the dict representation of the Backbone model 'CertificateException'
+
+    :param request:
+    :return: key-value pairs containing certificate exception data and User object
+    """
+    try:
+        certificate_exception = json.loads(request.body or '{}')
+    except ValueError:
+        raise ValueError(_('Invalid Json data'))
+
+    user = certificate_exception.get('user_name', '') or certificate_exception.get('user_email', '')
+    if not user:
+        raise ValueError(_('Student username/email is required.'))
+    try:
+        db_user = get_user_by_username_or_email(user)
+    except ObjectDoesNotExist:
+        raise ValueError(_('Student (username/email={user}) does not exist').format(user=user))
+    except MultipleObjectsReturned:
```

Comment:
`get_user_by_username_or_email` return a user object. is any test for this `MultipleObjectsReturned` scenario ?